Assets/UI/worldinput_CQUI.lua

```diff
@@ -0,0 +1,559 @@
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PASTURE"].Hash);
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PLANTATION"].Hash);
+        cquiHandledKey = true;
+      end
+    end
+  end
+
+  if (uiKey == Keys.R) then
+    if (unitType == "UNIT_MILITARY_ENGINEER") then
+      -- Build Road/Rail is a UnitOperation
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.UnitOperations["UNITOPERATION_BUILD_ROUTE"].Hash);
+      cquiHandledKey = true;
+    elseif (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.REBASE);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.S and CQUI_isAltDown == true) then
+    if (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestCommand(UI.GetHeadSelectedUnit(), UnitOperationTypes.AIR_ATTACK);
+      -- Use false as we need to let the unmodifed script deal with the Alt Key
+      cquiHandledKey =false;
+    end
+  end
+
+  if uiKey == Keys.VK_SHIFT then
+    CQUI_isShiftDown = false;
+  end
+
+  if uiKey == Keys.VK_ALT then
+    CQUI_isAltDown = false;
+  end
+
+  if (cquiHandledKey == false) then
+    print_debug("CQUI Did not handle key: "..tostring(uiKey));
+    cquiHandledKey = BASE_CQUI_DefaultKeyUpHandler( uiKey );
+  end
+
+  return cquiHandledKey;
+end
+
+-- ===========================================================================
+function OnDefaultKeyDown( pInputStruct:table )
+  print_debug("** Function Entry: OnDefaultKeyDown (CQUI Hook)");
+  if m_isInputBlocked then
+    print_debug("** m_isInputBlocked is true");
+    return;
+  end
+
+  return BASE_CQUI_OnDefaultKeyDown( pInputStruct );
+end
+
+-- ===========================================================================
+function OnDefaultKeyUp( pInputStruct:table )
+  print_debug("** Function Entry: OnDefaultKeyUp (CQUI Hook)");
+  if m_isInputBlocked then
+    print_debug("** m_isInputBlocked is true");
+    return;
+  end
+
+  return BASE_CQUI_OnDefaultKeyUp( pInputStruct );
+end
+
+-- ===========================================================================
+function OnPlacementKeyUp( pInputStruct:table )
+  print_debug("** Function Entry: OnPlacementKeyUp (CQUI Hook)");
+  if m_isInputBlocked then 
+    print_debug("** m_isInputBlocked is true");
+    return;
+  end
+
+  return BASE_CQUI_OnPlacementKeyUp( pInputStruct );
+end
+
+-- ===========================================================================
+function ClearAllCachedInputState()
+  print_debug("** Function Entry: ClearAllCachedInputState (CQUI Hook)");
+  BASE_CQUI_ClearAllCachedInputState();
+
+  CQUI_isShiftDown = false;
+end
+
+-- ===========================================================================
+function OnInputActionTriggered( actionId:number )
+ -- CQUI Enhanced Keyboard over-rides some of the ingame hot keys
+
```

Comment:
Is it safe and intentional to leave this function empty? The original seemed to only modify the part which handled previous city and next city.

Reply:
This is something I forgot to remove - I was messing with it when trying to figure out the issue with the arrow keys and what-not.  Removed it

Reply:
and this is totally why the "G" key was not being handled!  aaah!!! thank you for spotting this

Reply:
yep, left by mistake earlier.  Removed.  This was the issue preventing "G" from showing/hiding the grid!